framework/src/org/checkerframework/common/value/Range.java

```diff
@@ -0,0 +1,670 @@
+     * @return the range resulting from subtracting the specified range from this range
+     */
+    public Range minus(Range right) {
+        if (this.isNothing() || right.isNothing()) {
+            return NOTHING;
+        }
+
+        if (this.isWithinInteger() && right.isWithinInteger()) {
+            long resultFrom = from - right.to;
+            long resultTo = to - right.from;
+            return new Range(resultFrom, resultTo);
+        } else {
+            BigInteger bigFrom = BigInteger.valueOf(from).subtract(BigInteger.valueOf(right.to));
+            BigInteger bigTo = BigInteger.valueOf(to).subtract(BigInteger.valueOf(right.from));
+            return bigRange2LongRange(bigFrom, bigTo);
+        }
+    }
+
+    /**
+     * Returns the smallest range that includes all possible values resulting from multiplying an
+     * arbitrary value in the specified range by an arbitrary value in this range. We call this the
+     * multiplication of two ranges.
+     *
+     * @param right the specified range to be multiplied by this range
+     * @return the range resulting from multiplying the specified range by this range
+     */
+    public Range times(Range right) {
+        if (this.isNothing() || right.isNothing()) {
+            return NOTHING;
+        }
+
+        if (this.isWithinInteger() && right.isWithinInteger()) {
+            List<Long> possibleValues =
+                    Arrays.asList(
+                            from * right.from, from * right.to, to * right.from, to * right.to);
+            return new Range(Collections.min(possibleValues), Collections.max(possibleValues));
+        } else {
+            List<BigInteger> bigPossibleValues =
+                    Arrays.asList(
+                            BigInteger.valueOf(from).multiply(BigInteger.valueOf(right.from)),
+                            BigInteger.valueOf(from).multiply(BigInteger.valueOf(right.to)),
+                            BigInteger.valueOf(to).multiply(BigInteger.valueOf(right.from)),
+                            BigInteger.valueOf(to).multiply(BigInteger.valueOf(right.to)));
+            BigInteger bigFrom = Collections.min(bigPossibleValues);
+            BigInteger bigTo = Collections.max(bigPossibleValues);
+            return bigRange2LongRange(bigFrom, bigTo);
+        }
+    }
+
+    /**
+     * Returns the smallest range that includes all possible values resulting from dividing (integer
+     * division) an arbitrary value in this range by an arbitrary value in the specified range. We
+     * call this the division of two ranges.
+     *
+     * @param right the specified range by which this range is divided
+     * @return the range resulting from dividing this range by the specified range
+     */
+    public Range divide(Range right) {
+        if (this.isNothing() || right.isNothing()) {
+            return NOTHING;
+        }
+
+        if (right.contains(0)) {
+            // Should warn division by zero
```

Comment:
This warning would be overly conservative:  it would issue a warning at every division in unannotated code.  A different type-checker could focus on division-by-zero errors, but that is not the place of this one.